### PR TITLE
feat: Add SecurityContext to ContainerTemplate

### PR DESCRIPTION
**Pull request by samuel-hawker**
Added SecurityContext to ContainerTemplate and
added logic to handle it in all models which use it

Attached tests ensure that these values are then
propogated to deployments/sts

Also added a JmxTemplate and ContainerTemplate to
JmxTrans

Signed-off-by: Samuel Hawker <samuel.hawker@ibm.com>

### Type of change

_Select the type of your PR_

- Bugfix
- Enhancement / new feature
- Refactoring
- Documentation

### Description

_Please describe your pull request_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    private ContainerTemplate jmxTransContainer;
+    protected Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("Template for JmxTrans container")
+    @JsonInclude(JsonInclude.Include.NON_EMPTY)
+    public ContainerTemplate getJmxTransContainer() {
+        return jmxTransContainer;
+    }
+
+    public void setJmxTransContainer(ContainerTemplate tlsSidecarContainer) {
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java:41]: MAybe you could fix the variable name here? `tlsSidecarContainer` -> `jmxTransContainer`

File: api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.api.kafka.model;
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java:6]: Any reason why no to keep this in the `template` subpackage with the other template files?

File: api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        editableEnabled = false,
+        builderPackage = Constants.FABRIC8_KUBERNETES_API
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({ "jmxTransContainer", "additionalProperties" })
+@EqualsAndHashCode
+public class JmxTransTemplate implements Serializable, UnknownPropertyPreserving {
+    private static final long serialVersionUID = 1L;
+
+    private ContainerTemplate jmxTransContainer;
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java:32]: Since we are already adding it - it would be great to add also pod and deployment temlapate. We can do it in separate PR if you want - but since you are already adding this here might be easier to do everything.

File: api/src/main/java/io/strimzi/api/kafka/model/template/ContainerTemplate.java [hunk 4]
```diff
@@ -41,6 +43,16 @@ public void setEnv(List<ContainerEnvVar> env) {
         this.env = env;
     }
 
+    @Description("Security context for the container")
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/template/ContainerTemplate.java:47]: I guess it has some Kubernetes reference .... can we add the `@KubeLink` annotation to render it in the doc? It is done similarly in the `PodSecurityContext` in `PodTemplate`.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityOperatorTest.java [hunk 5]
```diff
@@ -732,4 +738,124 @@ public void testTlsSidecarContainerEnvVarsConflict() {
...
+                                .withSecurityContext(securityContext)
+                            .endTlsSidecarContainer()
+                        .endTemplate()
+                    .endEntityOperator()
+                .endSpec()
+                .build();
+
+        EntityOperator eo =  EntityOperator.fromCrd(resource, VERSIONS);
+
+        assertThat(eo.getTemplateTlsSidecarContainerSecurityContext(), is(securityContext));
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityOperatorTest.java:851]: Do we really need to test this? Is it not enough to test it in the pod template? It would allow us to remove the `getTemplateTlsSidecarContainerSecurityContext()` method. Just an idea, feel free to ignore if you think this is useful.

File: api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java (hunk 3 above)
Comment (scholzj) [inline api/src/main/java/io/strimzi/api/kafka/model/JmxTransTemplate.java:32]: I also wonder since this is. the main container in its pod if we should just call it `container` instead of `jmxTransContainer`. It is similar in the `KafkaExporterTemplate` for example.

**Comment by samuel-hawker** [comment]
FYI - in adding in the extra templates, I made a new custom matcher. I put it in ResourceUtils, I think that’s the right place to put it...

But it is importing quite a few classes, I wonder if its worth putting in a new class, something like `StrimziMatchers`

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java [hunk 6]
```diff
@@ -753,4 +758,32 @@ public static ClusterOperatorConfig dummyClusterOperatorConfig(long operationTim
     public static ClusterOperatorConfig dummyClusterOperatorConfig() {
         return dummyClusterOperatorConfig(new KafkaVersion.Lookup(emptyMap(), emptyMap(), emptyMap(), emptyMap(), emptyMap()));
     }
+
+    /**
+     * hasEntries is a custom matcher that checks is entries is a map whose entries
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/ResourceUtils.java:763]: Just a nitpick .. but this looks like some typos in the too many `is`.

**Comment by scholzj** [comment]
> FYI - in adding in the extra templates, I made a new custom matcher. I put it in ResourceUtils, I think that’s the right place to put it... But it is importing quite a few classes, I wonder if its worth putting in a new class, something like StrimziMatchers

I would probably create separate class ... but no strong opinions about this.

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by samuel-hawker** [comment]
Thanks for fixing it, my development environment appears to be erroring and not running the generation steps properly.

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by samuel-hawker** [comment]
Sure, no problem @ppatierno .
Just for clarification is this to retain an easier git history to read?

My IDE tends to sort and alphabetize my imports, I'll turn it off in future PRs.

**Comment by ppatierno** [comment]
@samuel-hawker thanks! Yes, it avoids having changes on files not related to the PR as well because your IDE changes the order.

**Comment by scholzj** [comment]
Thanks for the PR.
